### ISSUE TITLE
Add (simple) git support

### ISSUE THREAD

**Issue by koppor** [JabRef/jabref#12350]
As user, I would like to have a seamless integration in git. For "straight-forward" cases, JabRef should do "magic".

Implementation drivers

- In case of an error, the file should be in the state before JabRef started to do some wizard thing

## Opening a library

1. When opening a library (.bib file), JabRef checks whether the .bib file is contained in a git repository. If not in a git repository, no further action is done. Otherwise it is tagged as "versioned" (BibDatabaseContext)
2. JabRef exetues a "git pull" to update the .bib file
3. If error: notification of error, revert file to previous state
4. JabRef opens the updated file

## Saving of a library

No action taken (to save number of commits)

## Saving a library

1. JabRef saves the .bib file
2. A git commit is created. git commit message "Automatic update via JabRef".
3. ~~`git pull --rebase` is executed.~~ Just execute `git pull`
4. ~~In case of an error, revert to commit created at step 2 and try `git pull`~~
5. In case of an error, revert to commit created at step 2 and ouput a notification - and stop
6. Then,  `git push`
7. In case of an error, output error - and stop
8. Output "Successfully pushed"

---

## Advanced implementation

A user might want to trigger pushing

- Reuse button "synchronize" of shared database. This button has the same functionality as "Saving a library"
- Let user configure frequency of push. At this frequence, on "Saving a library", a commit and the steps of "Saving a library" are done.

---

## Implementation hints

- Working with JGit in general: https://github.com/koppor/jgit-mwe
- Git in JabRef: Start at org.jabref.logic.git.GitHandlerTest - and then read on at org.jabref.logic.git.GitHandler
- Please, start with http-based repositories: username + password. For GitHub, the password is a "personal access token". See https://github.com/koppor/jgit-mwe?tab=readme-ov-file#run-with-credentials.
- Some ideas can be taken from https://github.com/JabRef/jabref/pull/10586 (e.g., how to work with preferences)

**Comment by SihasA** [comment]
Hi, we would like to propose the following to solve this solution:

1) When a library is opened check whether it is being version controlled via functions in GitHandler.java provided in JabRef
2) Syncronise with remote repo as per preferences- automaticly/ manually. Current sync state of version controlled files are to be displayed appropriately.

![Image](https://github.com/user-attachments/assets/4a24263b-2a06-438e-a028-04a150e5880a)

![Image](https://github.com/user-attachments/assets/0817e870-3de1-4639-ac61-2f877f17318c)

3) Add a tab to preferences that handles "git preferences" managing optional auto push and sync. The exact preferences are to be further discussed and streamlined.

![Image](https://github.com/user-attachments/assets/02abf536-6950-41c0-b0e4-0bd7affd99b1)

4) Introduce a merge conflict resolution mode that introduces a GitHub style resolver with a side by side of the file, showcasing additions and deletions.

![Image](https://github.com/user-attachments/assets/7e7f1aa1-9ba0-4850-bf5c-20aed9363668)

5) We were also thinking of allowing users to access git pull/ commit/push commands and functionality by introducing either an individual field or a git tab under files.

![Image](https://github.com/user-attachments/assets/5ba6b207-005e-469f-a042-fcadc3f2229b)

Please let us know any thoughts on the following ideas and if this solves the proposed issue well enough.

**Comment by koppor** [comment]
Thank you for the initial proposal. I very much miss options to choose from.

Did you hear about the "Shared SQL Database" feature? --> https://docs.jabref.org/collaborative-work/sqldatabase

While reading your pseudo-flow-chart, I thought, integration with a git remote repo is something very similar. The "only" difference is that the library is saved locally, too. 

* Option A is now to implement a "Shared SQL Database" with "git" read as "Shared SQL Database" and add synchronization to a loal file. 
* Option B is to handle the synchroniaztion differently (as you described), but use the ideas of the "Shared SQL Database" feature. There, the thing of concurrent modifications is also thought of
* Option C is to implement everything new and ignore the "Shared SQL Database" knowledge at all.

I currently think, you proposal is doing Option C. I would ask to consider option B, please.

---


>     1. When a library is opened check whether it is being version controlled via functions in GitHandler.java provided in JabRef

Seems to be OK... It should be a property in ´org.jabref.model.database.BibDatabaseContext`. Similar to `bibDatabaseContext.getDatabasePath();`. See handling in `org.jabref.gui.LibraryTab#updateTabTitle` for unsaved libraries.

>     2. Syncronise with remote repo as per preferences- automaticly/ manually. Current sync state of version controlled files are to be displayed appropriately.

Yes

> ![Image](https://github.com/user-attachments/assets/4a24263b-2a06-438e-a028-04a150e5880a)

No. Use the tab title. See handling for unsaved libraries:

![Image](https://github.com/user-attachments/assets/1b5a2f8a-e3e7-451a-9648-1791bdb0f6e0)

JabRef stores in the tab title information on saved an unsaved. This is a property of the "state". The git sync state is also such a state. Thus, it should also be in the title.

You can use some unicode characters to display the state - to make the title short.


> ![Image](https://github.com/user-attachments/assets/0817e870-3de1-4639-ac61-2f877f17318c)

Proposal: First work on "Sync on save" - and then think of manual sync. - Otherwise, there is too much to think of.

**NO SVN SUPPORT** - git only (!)

There should be no "notification" in the sense of ´org.jabref.logic.util.NotificationService#notify`. Maybe rework this using the tab title and the ida of "sync-on-save"

>     3. Add a tab to preferences that handles "git preferences" managing optional auto push and sync. The exact preferences are to be further discussed and streamlined.

Add it to "General" as there is already a section "Saving"

![Image](https://github.com/user-attachments/assets/0711047e-7585-4c10-a117-6e8cd4a0bf6c)

You can also move "Saving" to a new tab "Saving" in the preference and add a sub section "Git" on that.

Please, do not mix up "commit" and "push". Please read some basic git tutorials to get the difference between "commit" and "push". Also learn about the staging area of git and the difference between "local" and "remote" repository. -- These terms should be used.

Use `JabRef auto check-in"  - it is based on A) what TeXStudio uses and B) how git writes "check-in": 

![Image](https://github.com/user-attachments/assets/296325e6-e4ac-4179-9978-89efe8ad4497)

> ![Image](https://github.com/user-attachments/assets/02abf536-6950-41c0-b0e4-0bd7affd99b1)
> 
>     4. Introduce a merge conflict resolution mode that introduces a GitHub style resolver with a side by side of the file, showcasing additions and deletions.

Here, two concepts are mixed:

1. Library diff
2. Conflict resolution

I very much like the concept of "Library diff". Very neat. This is a "viewer" for `org.jabref.logic.bibtex.comparator.BibDatabaseDiff`. This can be done as sub project (somehow) independent of the git support. -- The viewer could be extended with the functionality to pick and choose. Keep addition, keep deletion etc. **However**, the tool should merge as much as possible. Exmaple: I add entry number 999. The other user modifies entry 555. Then, the update of 555 should be in without any intervention of my side. Only if I modify 555 too, then there should be a dialog asking me which data I want to use. This should be done using the [Merge Entries Dialog](https://docs.jabref.org/finding-sorting-and-cleaning-entries/mergeentries dialog).

**However**, as first step, please use the existing functionality org.jabref.gui.collab.DatabaseChangesResolverDialog.


**Add test cases on BibTeX for things that can happen**. E.g., 

- two parties modify the same BibEntry, 
- one party adds an entry, another one removes it, but they are very far away and the can be merged without conflict.

>     5. We were also thinking of allowing users to access git pull/ commit/push commands and functionality by introducing either an individual field or a git tab under files.

This is OK. For the manual mode (see above). -- Should be 3 to hours of work.

Please think more in the BibTeX domain. How to uniquly identify a BibEntry, different kind of "merge conflicts". `org.jabref.logic.bibtex.comparator.BibDatabaseDiff#compareEntries` is already a good start. See also the test cases.

> Please let us know any thoughts on the following ideas and if this solves the proposed issue well enough.

This is a good start, but much more work ahead. Maybe, you can split up and have persons working on test cases.

I did not see any back reference to comments in this issue. However, I think, the really important was to pointer to the existing database conflict resolution dialog.

Please note that there was an existing try at https://github.com/JabRef/jabref/pull/10586. That was OKish - **However**, do not force the user to configure the git repository. You can assume an existing one and a proper git configuration (username + email). -- I just remembered the PR, because it had much code because of SSH - and it could be that this is necessary, but I am not sure.

As advanced step, you can allow for creation of a local git repository and configuraiton of the remote. However, this could be too much effort (also w.rt. maintenance).

---

Please also think about test cases: Automatic interacton with another git repository. I assume, it has to be made available as sub module in JabRef's code?

Code hitns (for another project): https://github.com/koppor/heylogs/blob/34c84ebb101b310bd0d715c669ab38e6d0300e65/heylogs-api/src/test/java/internal/heylogs/GitDiffRuleTest.java#L120

---

Related Work:

### TeXStudio

Configuration:

![Image](https://github.com/user-attachments/assets/99eb4321-e2d6-42c8-b045-f4680b49a4ad)

The two yellow options should be offered by JabRef, too.

The "to undo before last saved operation" is very interesting.

![Image](https://github.com/user-attachments/assets/22042135-7721-48c3-a7c0-c2ebdd2b4b35)

This means: As user, I have git in place. I close JabRef. I open JabRef. I press "Undo". Normally, there would be no Undo information. However, with "git", there is undo possible.

Manual possibilities:

![Image](https://github.com/user-attachments/assets/83cad9fd-b3df-40bb-953c-e1e847b87c4b)

**Comment by koppor** [comment]
One can experience the current file-based conflict handling as follows:

1. Open "Chocolate.bib" in JabRef. (The file is located in src\test\resources\testbib). 
2. Open "Chocolate.bib" in "Notepad++" 
3. Delete the BibEntry "Keen_2001".
4. See in JabRef\
    ![Image](https://github.com/user-attachments/assets/f0dcef20-ce0f-4df5-9a3c-e6f5c555e28b)
5. Click on "Review Changes" opens a dialog showing kind of a diff between the local file and the in-memory state in JabRef:\
    ![Image](https://github.com/user-attachments/assets/9b1e4875-af2f-4e0c-9e88-cda39e622a97)

**Comment by koppor** [comment]
Notes on discussion:

- On push: If it does not work: pull. If these are conflicts: DO NOT force the user to resolve the conflicts. Let the user postpone the merge conflict resolution (and let him continue to work)
- "git commit --ammend" vs. "git commit" (maybe in the case non-auto-sync to auto-sync) -- this is simlar to choose MariaDB or Postgres as remote SQL database (the first one uses manual sync - and the last one uses auto sync)
- Show branch in status (tab)
- (some more internal notes exist)

**Comment by LinusDietz** [comment]
Thanks! I think if you work in parallel, you should be able to tackle WP1 and WP2 until Friday.

**Comment by SihasA** [comment]
Alright perfect! Will start on both ASAP

**Comment by bauer-at-work** [comment]
Nice to see that git support is being worked on.

In the meantime, fyi, I use the following systemd user service to automatically my changes:

```ini
[Unit]
Description=git auto-commit&push on changing work.bib
After=network.target

[Service]
ExecStart=%h/.local/share/cargo/bin/watchexec --workdir=%h/repos/bib -e bib --postpone -- git commit -am \"Automatic update\" && git push
Restart=on-failure

[Install]
WantedBy=default.target
```

**Comment by Siedlerchr** [comment]
Related https://github.com/JabRef/jabref/pull/12252

**Comment by subhramit** [comment]
@wanling0000 could you comment on this issue so that you can be assigned

**Comment by wanling0000** [comment]
I would like to work on this issue
/assign-me

**Comment by hasibagen** [comment]
I'd like to ask about the current progress. I've read these links, but they don't seem to have much to do with Git support?

If I want to put a JabRef directory on Git for multi-user synchronization, can I use it directly?

The new shared database seems to require SQL, but I just want to use Git.

https://github.com/JabRef/jabref/pull/5669
https://github.com/JabRef/jabref/pull/14102
https://github.com/JabRef/jabref/blob/main/docs/code-howtos/remote-storage-jabdrive.md

JabRef 5.15--2024-07-10--1eb3493
Linux 6.17.0-22-generic amd64
Java 21.0.2
JavaFX 22.0.1+7
My computer version

**Comment by wanling0000** [comment]
> I'd like to ask about the current progress. I've read these links, but they don't seem to have much to do with Git support?
> 
> If I want to put a JabRef directory on Git for multi-user synchronization, can I use it directly?
> 
> The new shared database seems to require SQL, but I just want to use Git.
> 
> [#5669](https://github.com/JabRef/jabref/pull/5669) [#14102](https://github.com/JabRef/jabref/pull/14102) https://github.com/JabRef/jabref/blob/main/docs/code-howtos/remote-storage-jabdrive.md
> 
> JabRef 5.15--2024-07-10--1eb3493 Linux 6.17.0-22-generic amd64 Java 21.0.2 JavaFX 22.0.1+7 My computer version

Hi, thanks for your question.

We introduced an experimental Git integration in JabRef. You can find an overview here:
https://blog.jabref.org/2025/11/11/introducing-git-integration-in-jabref/

It is still in an early stage and not yet a fully mature replacement for real-time collaboration (like the shared SQL database).  Could you take a look and see if this matches your use case for Git-based collaboration?

If something important is missing or doesn’t work as expected, feel free to open an issue to help us improve the feature.
